crates/protocol/src/operator/validate.rs

```diff
@@ -113,31 +113,29 @@ impl Validator {
     /// fails to validate, the validator state will remain unchanged.
     pub fn validate(
         &mut self,
-        envelope: &ProtoEnvelope<model::OperatorRecord>,
+        record: &ProtoEnvelope<model::OperatorRecord>,
     ) -> Result<(), ValidationError> {
         let snapshot = self.snapshot();
 
-        let result = self.validate_envelope(envelope);
+        let result = unsafe { self.validate_record(record) };
         if result.is_err() {
             self.rollback(snapshot);
         }
 
         result
     }
 
-    /// Gets the public key of the given key id.
+    /// Validates an individual operator record without a transaction.
     ///
-    /// Returns `None` if the key id is not recognized.
-    pub fn public_key(&self, key_id: &signing::KeyID) -> Option<&signing::PublicKey> {
-        self.keys.get(key_id)
-    }
-
-    fn initialized(&self) -> bool {
-        // The package log is initialized if the hash algorithm is set
-        self.algorithm.is_some()
-    }
-
-    fn validate_envelope(
+    /// # Safety
+    ///
+    /// This is marked as unsafe as the caller must ensure that the validator
+    /// is snapshotted before calling this method and rolled back if the
+    /// validation fails.
+    ///
+    /// Failure to do so may result in the validator being left in an inconsistent
+    /// state.
+    pub unsafe fn validate_record(
```

Comment:
IMO either this isn't actually `unsafe` or we need to change it. A basic API like this shouldn't be designed in a way that needs `unsafe`.

Reply:
I think I can remove this method; it was introduced early on with these changes as a way to prevent repeated validator snapshotting, but the remaining place we do an explicit `snapshot` call isn't necessary any more.